### PR TITLE
Tests, refactor and coverage on package opts

### PR DESCRIPTION
**Pull request by vdemeester**
This is the time for `opts` to get covered 🐸 and a little bit refactored.

`ValidateEnv` and `ParseEnvFile` uses a regexp to valide that it is a correct environment variable that is passed. The definition used : 

> Variables set by the user must have a name consisting solely of alphabetics, numerics, and underscores - the first of which must not be numeric.

`ValidateLabel` is made consistent with the `LABEL` in `Dockerfile` as it allows multiple `=` like `LABEL label=value1=value2`.

``` bash
# before
+ go test -test.timeout=30m github.com/docker/docker/opts
PASS
coverage: 33.8% of statements
#after
+ go test -test.timeout=30m github.com/docker/docker/opts
PASS
coverage: 92.8% of statements
```

There is a few `FIXME` in this one :
- <del>`ValidatePath(":/test")` is valid, although it will fail after I think</del>
- <del>`func ValidateEnv(val string) (string, error)` always returns `nil` for error. I guess it's for _coherence of returns_ with the other validate method, so I think I can remove my `FIXME` comment.</del>
- <del>There is `FIXMEs` for `ListOpts.GetAll()` and `ListOpts.GetMap` asking if we can remove it. They are used in `runconfig/parse.go` ; `GetMap` could/should probably have another name but..</del>

Signed-off-by: Vincent Demeester vincent@sbr.pm

### CODE REVIEW

File: opts/opts.go [hunk 1]
```diff
@@ -215,6 +215,7 @@ func ValidatePath(val string) (string, error) {
 	return val, nil
 }
 
+// FIXME should it return error as it is always nil ?
```
Comment (calavera) [inline opts/opts.go]: why don't you fix it rather than leaving a comment? :wink:

Comment (LK4D4) [inline opts/opts.go]: Yup, and it shouldn't :)

Comment (vdemeester) [inline opts/opts.go]: Well, I was asking if this was ok before doing it :smile_cat:. Now that I know it is I will fix it :wink:.

Comment (vdemeester) [inline opts/opts.go]: Ok that's kinda what I guessed, it has to be there because it is used as a ``type ValidatorFctType func(val string) (string, error)``. I think I should just remove the comment for now :stuck_out_tongue_closed_eyes:.

Environment and thus Label (that are validated the same way) are always *valid*, just rewritten a bit sometimes.

Comment (LK4D4) [inline opts/opts.go]: Haha :)

File: opts/envfile_test.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	defer os.RemoveAll(f.Name())
+	testFile := []byte(`first line
+another invalid line`)
+	if _, err := f.Write(testFile); err != nil {
+		t.Fatal(err)
+	}
+
+	envs, err := ParseEnvFile(f.Name())
+
+	if err != nil {
```
Comment (duglin) [inline opts/envfile_test.go]: IMO we're inconsistent. If you added a "=" after each of those you'd get an error like:
```
docker: poorly formatted environment: variable 'a b c' has white spaces.
```
Yet w/o the "=" it passes it thru to the container. We should either allow both or error both.
I think we should detect the spaces and error in both cases.  Want to fix that as part of this?

Comment (vdemeester) [inline opts/envfile_test.go]: @duglin I sure do :wink:

File: opts/ip_test.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	"testing"
+)
+
+func TestNewIpOptInvalidDefaultVal(t *testing.T) {
+	ip := net.IPv4(127, 0, 0, 1)
+	defaultVal := "Not an ip"
+
+	ipOpt := NewIpOpt(&ip, defaultVal)
+
+	expected := "127.0.0.1"
```
Comment (duglin) [inline opts/ip_test.go]: Kind of bummed, and weird, that this isn't an error case.

File: opts/ip_test.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	if ipOpt.String() != expected {
+		t.Fatalf("Expected [%v], got [%v]", expected, ipOpt.String())
+	}
+}
+
+func TestNewIpOptValidDefaultVal(t *testing.T) {
+	ip := net.IPv4(127, 0, 0, 1)
+	defaultVal := "192.168.1.1"
+
+	ipOpt := NewIpOpt(&ip, defaultVal)
```
Comment (duglin) [inline opts/ip_test.go]: not really a comment on the test, but I just don't get all of the levels of indirection in this IpOpt stuff.  Maybe I'm missing something but why not just have it be `NewIpOpt(defaultValue)`??  What's the point of passing in the 'ip' if all we're going to do is try to override it with 'defaultValue' ?  @LK4D4 do you know the history here?

File: opts/envfile_test.go (hunk 2 above)
Comment (vdemeester) [inline opts/envfile_test.go]: There is a *certain* consistency in the fact that we are able to run ``docker run -e "something with space" busybox /bin/sh`` ; although the environment variables is not set in the container. I should probably validate this too (``ValidateEnv``).

File: opts/opts_test.go [hunk 5]
```diff
@@ -85,8 +90,44 @@ func TestListOpts(t *testing.T) {
...
+	if len(mapListOpts) != 1 {
+		t.Errorf("Expected [map[bar:{}]], got [%v]", mapListOpts)
+	}
+
+}
+
+func TestListOptsWithValidator(t *testing.T) {
+	// Re-using logOptsvalidator (used by MapOpts)
+	o := NewListOpts(logOptsValidator)
+	o.Set("foo")
```
Comment (duglin) [inline opts/opts_test.go]: Any reason you're not also checking to see that o.Set() returns an err ?

Comment (duglin) [inline opts/opts_test.go]: eh, never mind - same net result and I guess the there's only one type of error that can be returned so no need to check.

File: opts/opts_test.go [hunk 6]
```diff
@@ -179,6 +220,163 @@ func TestValidateExtraHosts(t *testing.T) {
 	}
 }
 
+func TestValidateAttach(t *testing.T) {
+	valid := []string{
+		"stdin",
+		"stdout",
+		"stderr",
```
Comment (duglin) [inline opts/opts_test.go]: You should probably add some non-lower case values here to make sure we lower-case them in the ValidateAttach return value.

File: opts/opts_test.go [hunk 7]
```diff
@@ -179,6 +220,163 @@ func TestValidateExtraHosts(t *testing.T) {
...
+}
+
+func TestValidatePath(t *testing.T) {
+	valid := []string{
+		"/home",
+		"/home:/home",
+		"/home:/something/else",
+		"/with space",
+		"/home:/with space",
+		":/test", // FIXME should probably fail
```
Comment (duglin) [inline opts/opts_test.go]: I agree - let's fix this so it fails.

Comment (duglin) [inline opts/opts_test.go]: let's add "relative-path:abs-path" so that we make sure its only checking that container-path is absolute.

File: opts/opts_test.go [hunk 8]
```diff
@@ -179,6 +220,163 @@ func TestValidateExtraHosts(t *testing.T) {
...
+		"":        " is not an absolute path",
+		"./":      "./ is not an absolute path",
+		"../":     "../ is not an absolute path",
+		"/:../":   "../ is not an absolute path",
+		"/:path":  "path is not an absolute path",
+		":":       " is not an absolute path",
+		"/tmp:":   " is not an absolute path",
+		":test":   "test is not an absolute path",
+		"tmp:":    " is not an absolute path",
+		":test:":  "test: is not an absolute path",
```
Comment (duglin) [inline opts/opts_test.go]: just to be anal, since there are other edits you need to make, let's add `::` to the list.

File: opts/opts_test.go [hunk 9]
```diff
@@ -179,6 +220,163 @@ func TestValidateExtraHosts(t *testing.T) {
...
+	}
+}
+
+func TestValidateLabel(t *testing.T) {
+	if _, err := ValidateLabel("label"); err == nil || err.Error() != "bad attribute format: label" {
+		t.Fatalf("Expected an error [bad attribute format: label], go %v", err)
+	}
+	if actual, err := ValidateLabel("key1=value1"); err != nil || actual != "key1=value1" {
+		t.Fatalf("Expected [key1=value1], got [%v,%v]", actual, err)
+	}
```
Comment (duglin) [inline opts/opts_test.go]: let's make sure that "key1=value1=value2" passes too.

Comment (vdemeester) [inline opts/opts_test.go]: Actually it won't pass and will return ``bad attribut format …``, because in the ``ValidateLabel`` it checks for only one ``=``.

File: opts/opts_test.go (hunk 6 above)
Comment (vdemeester) [inline opts/opts_test.go]: Done :wink:

File: opts/opts_test.go (hunk 9 above)
Comment (duglin) [inline opts/opts_test.go]: interesting.  You can do `LABEL foo=bar=car` in a Dockerfile.  Seems we're a bit inconsistent.
@rhatdan any reason we ban "=" in the value part in the Validate func?

Comment (vdemeester) [inline opts/opts_test.go]: Ah well, I didn't check that on the ``LABEL`` for the ``Dockerfile``.. yeah seems a bit inconsistent.

**Comment by vdemeester** [comment]
Ok, so I've updated the ``envfile`` end ``env`` validation.

- For ``envfile``, it will fail if there is spaces even without ``=``, a file like this would fail :

  ```bash
  some spaces
  ```

- Almost the same for ``env`` (and now there is a case it returns an error :stuck_out_tongue_closed_eyes: ), it's no more possible to do something like : ``docker run -e "some spaces" …``. 

For the sake of it, I wasn't sure if it was possible to define a environment variable with spaces, so I tried (with ``bash`` and ``zsh``) : 
```bash
$ export SOME\ SPACES=value
bash: export: `SOME SPACES=value': not a valid identifier
$ export "SOME SPACES"=value
bash: export: `SOME SPACES=value': not a valid identifier
```

:relaxed:

**Comment by duglin** [comment]
yep - that's what I got in my testing as well.  thanks!!

**Comment by rhatdan** [comment]
LABEL call just stole code from ENV.  It should follow the same standard of whatever ENV allows, I believe.

**Comment by vdemeester** [comment]
Rebased to master and move a test from https://github.com/docker/docker/pull/13879 in ``ip_test.go``.

File: opts/envfile.go [hunk 10]
```diff
@@ -23,14 +23,14 @@ func ParseEnvFile(filename string) ([]string, error) {
...
-				variable := strings.TrimLeft(data[0], whiteSpaces)
-				if strings.ContainsAny(variable, whiteSpaces) {
-					return []string{}, ErrBadEnvVariable{fmt.Sprintf("variable '%s' has white spaces", variable)}
-				}
+			// trim the front of a variable, but nothing else
+			variable := strings.TrimLeft(data[0], whiteSpaces)
+			if strings.ContainsAny(variable, whiteSpaces) {
+				return []string{}, ErrBadEnvVariable{fmt.Sprintf("variable '%s' has white spaces", variable)}
+			}
+			if strings.Contains(line, "=") {
```
Comment (duglin) [inline opts/envfile.go]: minor thing, and up to you, but I think `if len(data) > 1` would be better just because it'll be faster than looping over the string looking for `=`

File: opts/ip_test.go [hunk 11]
```diff
@@ -0,0 +1,54 @@
+package opts
+
+import (
+	"net"
+	"testing"
+)
+
+func TestIpOptString(t *testing.T) {
```
Comment (duglin) [inline opts/ip_test.go]: I think this is covered by #13879.  hmmm did they steal your code?

File: opts/opts_test.go [hunk 12]
```diff
@@ -180,14 +220,182 @@ func TestValidateExtraHosts(t *testing.T) {
...
+		"angry_torvalds:linus",
+	}
+	invalid := map[string]string{
+		"":               "empty string specified for links",
+		"too:much:of:it": "bad format for links: too:much:of:it",
+	}
+
+	for _, link := range valid {
+		if _, err := ValidateLink(link); err != nil {
+			t.Fatalf("ValidateLink(`%q`) should succeed: erro %v", link, err)
```
Comment (duglin) [inline opts/opts_test.go]: s/erro/error/

Comment (duglin) [inline opts/opts_test.go]: s/%v/%q so we add quotes

File: opts/opts_test.go [hunk 13]
```diff
@@ -180,14 +220,182 @@ func TestValidateExtraHosts(t *testing.T) {
...
+		":test:":  "bad format for volumes: :test:",
+		"::":      "bad format for volumes: ::",
+		":::":     "bad format for volumes: :::",
+		"/tmp:::": "bad format for volumes: /tmp:::",
+		":/tmp::": "bad format for volumes: :/tmp::",
+	}
+
+	for _, path := range valid {
+		if _, err := ValidatePath(path); err != nil {
+			t.Fatalf("ValidatePath(`%q`) should succeed: erro %v", path, err)
```
Comment (duglin) [inline opts/opts_test.go]: s/erro/error/

Comment (duglin) [inline opts/opts_test.go]: %v->%q

File: opts/opts_test.go [hunk 14]
```diff
@@ -180,14 +220,182 @@ func TestValidateExtraHosts(t *testing.T) {
...
+}
+
+func TestValidateLabel(t *testing.T) {
+	if _, err := ValidateLabel("label"); err == nil || err.Error() != "bad attribute format: label" {
+		t.Fatalf("Expected an error [bad attribute format: label], go %v", err)
+	}
+	if actual, err := ValidateLabel("key1=value1"); err != nil || actual != "key1=value1" {
+		t.Fatalf("Expected [key1=value1], got [%v,%v]", actual, err)
+	}
+	if _, err := ValidateLabel("key1=value1=value2"); err == nil || err.Error() != "bad attribute format: key1=value1=value2" {
```
Comment (duglin) [inline opts/opts_test.go]: I thought we were going to allow "=" in the label value? I can do:
```
docker run -d --label=foo=abc=def ubuntu sleep 60
```
and it works just fine.

File: opts/ulimit_test.go [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+func TestUlimitOpt(t *testing.T) {
+	ulimitMap := map[string]*ulimit.Ulimit{
+		"nofile": {"nofile", 1024, 512},
+	}
+
+	ulimitOpt := NewUlimitOpt(ulimitMap)
+
+	expected := "[nofile=512:1024]"
+	if ulimitOpt.String() != expected {
+		t.Fatalf("Expected [%v], got [%v]", expected, ulimitOpt)
```
Comment (duglin) [inline opts/ulimit_test.go]: nit but any reason for the double [ ]'s  for 'expected'?

File: opts/ip_test.go (hunk 11 above)
Comment (vdemeester) [inline opts/ip_test.go]: Yes it is covered by #13879, I've just rebased against and moved it into this file ; so no if anybody *stole* it, it's me :sweat_smile:.

File: opts/envfile.go (hunk 10 above)
Comment (vdemeester) [inline opts/envfile.go]: sounds about right, will do :)

File: opts/ulimit_test.go (hunk 15 above)
Comment (vdemeester) [inline opts/ulimit_test.go]: not really :sweat_smile:, I'll update that too :)

File: opts/opts_test.go (hunk 14 above)
Comment (vdemeester) [inline opts/opts_test.go]: Oh right.. did not updated that. I will :wink:

**Comment by vdemeester** [comment]
Any more thoughts one that ? :blush:

File: opts/envfile.go [hunk 16]
```diff
@@ -23,14 +23,14 @@ func ParseEnvFile(filename string) ([]string, error) {
...
-				// trim the front of a variable, but nothing else
-				variable := strings.TrimLeft(data[0], whiteSpaces)
-				if strings.ContainsAny(variable, whiteSpaces) {
-					return []string{}, ErrBadEnvVariable{fmt.Sprintf("variable '%s' has white spaces", variable)}
-				}
+			data := strings.SplitN(line, "=", 2)
+
+			// trim the front of a variable, but nothing else
+			variable := strings.TrimLeft(data[0], whiteSpaces)
+			if strings.ContainsAny(variable, whiteSpaces) {
```
Comment (duglin) [inline opts/envfile.go]: I would think it would be better to use a regex and validate the env var name against the sh definition:
Variables set by the user must have a name consisting solely of alphabetics, numerics, and underscores - the first of which must not be numeric.
I know this is old code but I noticed it another PR and we might as well fix it right.

Comment (vdemeester) [inline opts/envfile.go]: Ok, sounds good to me :) I'll look at this real soon :).

File: opts/envfile_test.go [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+// ParseEnvFile with an invalid file
+func TestParseEnvFileInvalidFile(t *testing.T) {
+	f, err := ioutil.TempFile("", "tail-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer f.Close()
+	defer os.RemoveAll(f.Name())
```
Comment (duglin) [inline opts/envfile_test.go]: Minor thing but I think these defer's need to be switched so that we close it before we remove it. Defers are done in LIFO order.

File: opts/envfile_test.go [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+// ParseEnvFile with a random file, pass through
+func TestParseEnvFileRandomFile(t *testing.T) {
+	f, err := ioutil.TempFile("", "tail-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer f.Close()
+	defer os.RemoveAll(f.Name())
```
Comment (duglin) [inline opts/envfile_test.go]: swap order of defer's

File: opts/opts.go [hunk 19]
```diff
@@ -222,6 +225,9 @@ func ValidateEnv(val string) (string, error) {
 	if len(arr) > 1 {
 		return val, nil
 	}
+	if strings.ContainsAny(val, whiteSpaces) {
```
Comment (duglin) [inline opts/opts.go]: See comment above about using a regex

File: opts/envfile_test.go [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+// ParseEnvFile with a valid, well formated line
+func TestParseEnvFile(t *testing.T) {
+	f, err := ioutil.TempFile("", "tail-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer f.Close()
+	defer os.RemoveAll(f.Name())
```
Comment (duglin) [inline opts/envfile_test.go]: swap

File: opts/opts.go (hunk 19 above)
Comment (duglin) [inline opts/opts.go]: Odd that we only check for whitespace when there's a "=" in 'val'.
Seems like we should check for the validity of the chars of the variable name in all cases by doing this if-statement (as a regex) irrespective of whether there's a "=" in there or not.

File: opts/opts.go [hunk 21]
```diff
@@ -277,7 +283,7 @@ func ValidateExtraHost(val string) (string, error) {
 }
 
 func ValidateLabel(val string) (string, error) {
-	if strings.Count(val, "=") != 1 {
+	if strings.Count(val, "=") < 1 {
 		return "", fmt.Errorf("bad attribute format: %s", val)
 	}
```
Comment (duglin) [inline opts/opts.go]: @rhatdan are there any constraints on the valid chars for a LABEL name? Same as for env vars?

File: opts/opts_test.go [hunk 22]
```diff
@@ -180,14 +220,187 @@ func TestValidateExtraHosts(t *testing.T) {
...
+}
+
+func TestValidatePath(t *testing.T) {
+	valid := []string{
+		"/home",
+		"/home:/home",
+		"/home:/something/else",
+		"/with space",
+		"/home:/with space",
+		"retalive:/absolute-path",
```
Comment (duglin) [inline opts/opts_test.go]: s/retalive/relative/   unless you really wanted to mess with people's heads  :-)

File: opts/opts_test.go [hunk 23]
```diff
@@ -180,14 +220,187 @@ func TestValidateExtraHosts(t *testing.T) {
...
+			if err.Error() != expectedError {
+				t.Fatalf("ValidatePath(`%q`) error should contain %q, got %q", path, expectedError, err.Error())
+			}
+		}
+	}
+}
+
+func TestValidateEnv(t *testing.T) {
+	invalids := map[string]string{
+		"some spaces": "poorly formatted environment: variable 'some spaces' has white spaces",
```
Comment (duglin) [inline opts/opts_test.go]: once we get the regex stuff in there we should add some other invalid names, like:
```
asd!qwe
1asd
123
```

File: opts/opts_test.go [hunk 24]
```diff
@@ -180,14 +220,187 @@ func TestValidateExtraHosts(t *testing.T) {
...
+	}
+	valids := map[string]string{
+		"something":          "something",
+		"env1=value1":        "env1=value1",
+		"env2=value2=value3": "env2=value2=value3",
```
Comment (duglin) [inline opts/opts_test.go]: add:
```
env3=abc!qwe
```
just to make sure the regex we add doesn't bleed into the value part.  The = check should cover it, but since we're splitting on ='s I figured we should have a test with some other char.

File: opts/ulimit_test.go [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+	if err := ulimitOpt.Set("core=1024:1024"); err != nil {
+		t.Fatal(err)
+	}
+
+	// Invalid ulimit type returns an error and do not append to opts
+	if err := ulimitOpt.Set("notavalidtype=1024:1024"); err == nil {
+		t.Fatalf("Expected error on invalid ulimit type")
+	}
+	expected = "[nofile=512:1024 core=1024:1024]"
+	if ulimitOpt.String() != expected {
```
Comment (duglin) [inline opts/ulimit_test.go]: we may not get the same results each time since, I think, the order in which 'go' will iterate over maps is non-deterministic. it might be safer to make this:
```
expected2 = "[core=1024:1024 nofile=512:1024]"
result = ulimitOpt.String()
if result != expected && result != expected2 {
```

File: opts/opts_test.go (hunk 22 above)
Comment (vdemeester) [inline opts/opts_test.go]: At least my head was messed up when I wrote this one :sweat_smile:

File: opts/opts.go (hunk 21 above)
Comment (thaJeztah) [inline opts/opts.go]: It was discussed to add a `[a-zA-Z0-9][a-zA-Z0-9_.-]` restriction (https://github.com/docker/docker/pull/9882#issuecomment-70106777) but that didn't make it in the PR (see https://github.com/docker/docker/pull/9882#issuecomment-73987758) apart from being [documented as "should consist of"](https://github.com/docker/docker/pull/9882/files#diff-7dc7275791c2e807567d5e8a89d3f53bR44).

(I would still be in favour of adding that restriction, but that's out of scope here)

Comment (vdemeester) [inline opts/opts.go]: @thaJeztah Ok so it's not the same regexp as an environment variable. I think it makes more sense to not include this restriction here for the ``LABEL`` then.

Comment (thaJeztah) [inline opts/opts.go]: Unless the regex was added at a later stage (I'm not aware of that, but I might have missed it) :smile:

Comment (vdemeester) [inline opts/opts.go]: @thaJeztah I don't think it was added :stuck_out_tongue:. I should have said "the expression is not the same" ; meaning it's not gonna be the same regexp to use anyway :smile:

Comment (thaJeztah) [inline opts/opts.go]: Ah, thanks (sorry, nipping in and out of issues here :-))

Comment (rhatdan) [inline opts/opts.go]: I see no reason to restrict the label name.  Although Spaces always cause problems.   :^(

**Comment by vdemeester** [comment]
Janky is not gonna build it today :smile_cat:.

@duglin did the ``regexp`` stuff (I've put it in the ``envfile.go`` because it seemed like the right spot, but I saw later a bunch of regexp were in ``opts.go`` so I'm wondering it that was the right move or not).

File: opts/ip_test.go (hunk 11 above)
Comment (girliechicago) [inline opts/ip_test.go]: integration and opensource project is about collaborating with other Dev teams or the likes. There is no point of stealing a code that is being work on... Progression is real and not a vitual visibility that just happens by chance and we are not reboots....

**Comment by thaJeztah** [comment]
Tests are green!

@duglin @calavera @LK4D4 @rhatdan PTAL

**Comment by vdemeester** [comment]
Updated and rebased with the changes introduced by https://github.com/docker/docker/pull/14321.

**Comment by LK4D4** [comment]
LGTM
ping @duglin

File: opts/envfile.go [hunk 26]
```diff
@@ -4,9 +4,16 @@ import (
...
 	"fmt"
 	"os"
+	"regexp"
 	"strings"
 )
 
+var (
+	// Variables set by the user must have a name consisting solely of
+	// alphabetics, numerics, and underscores - the first of which must not be numeric.
+	EnvironmentVariableRegexp = regexp.MustCompile("^[[:alpha:]][[:alpha:][:digit:]_]+$")
```
Comment (duglin) [inline opts/envfile.go]: I think we need to modify this so that underscore is a valid first char.

File: opts/envfile_test.go [hunk 27]
```diff
@@ -84,15 +81,19 @@ func TestParseEnvFileBadlyFormattedFile(t *testing.T) {
     f   =quux
```
Comment (duglin) [inline opts/envfile_test.go]: @vdemeester any chance we can also test for other bad env names, like ones that start with a digit, and ones that have non-alphanumeric(_) chars in them (aside from spaces) - like "@" ?

File: opts/opts.go [hunk 28]
```diff
@@ -203,6 +203,9 @@ func ValidatePath(val string) (string, error) {
...
 
 	splited := strings.SplitN(val, ":", 2)
+	if splited[0] == "" {
+		return val, fmt.Errorf("bad format for volumes: %s", val)
+	}
 	if len(splited) == 1 {
 		containerPath = splited[0]
 		val = path.Clean(splited[0])
```
Comment (duglin) [inline opts/opts.go]: I'm pretty sure the logic in the 'else' part of this doesn't work when the input is `path:rw` because containerPath will be set to 'rw'.  The new testcases added doesn't test for this case, can we add it to see if I'm correct?

Comment (duglin) [inline opts/opts.go]: To be clear, I think the parsing of this is always going to be a bit ambiguous because you can never really tell if `rw:rw` is hostPath:containerPath or containerPath:rw.  I would suggest that when we only have one ":" we look at the 2nd part and if its `rw` or `ro` then we assume the first part is a containerPath.  Anything other than `rw` and `ro` means we have hostPath:containerPath.  Not foolproof, but we have no other way of knowing what the user meant.

File: opts/opts_test.go [hunk 29]
```diff
@@ -180,14 +220,192 @@ func TestValidateExtraHosts(t *testing.T) {
...
+}
+
+func TestValidatePath(t *testing.T) {
+	valid := []string{
+		"/home",
+		"/home:/home",
+		"/home:/something/else",
+		"/with space",
+		"/home:/with space",
+		"relative:/absolute-path",
```
Comment (duglin) [inline opts/opts_test.go]: Add some tests like:
```
/path:rw   # should work
path:ro   # should fail because path isn't abs
/rw:rw   # should work with /rw as the container path
/rw:/ro   # should work with /rw as the host path and /ro as the container path
```

File: opts/envfile_test.go (hunk 27 above)
Comment (duglin) [inline opts/envfile_test.go]: never mind - I see them in TestValidateEnv()

File: opts/opts_test.go [hunk 30]
```diff
@@ -180,14 +220,192 @@ func TestValidateExtraHosts(t *testing.T) {
...
+		"123":         "poorly formatted environment: variable '123' is not a valid environment variable",
+	}
+	valids := map[string]string{
+		"something":          "something",
+		"env1=value1":        "env1=value1",
+		"env2=value2=value3": "env2=value2=value3",
+		"env3=abc!qwe":       "env3=abc!qwe",
+		"env_4=value 4":      "env_4=value 4",
+		"PATH":               fmt.Sprintf("PATH=%v", os.Getenv("PATH")),
+		"PATH=something":     "PATH=something",
```
Comment (duglin) [inline opts/opts_test.go]: Add a test with underscope as the first char - it should pass.

File: opts/envfile.go (hunk 26 above)
Comment (vdemeester) [inline opts/envfile.go]: @duglin ohh right, I assumed it had to start with a *alpha* char, but reading back the comment I put, yep I missed the case of a variable starting by ``_``. I'll update it.

File: opts/opts.go (hunk 28 above)
Comment (vdemeester) [inline opts/opts.go]: @duglin sounds reasonable, I'll give it a try.

I have a question though, about those modes : https://github.com/docker/docker/blob/53d9609de472f2fc9e6aaef2b713f526605f4ab8/daemon/volumes.go#L115. We should handle them as well this way too.

Comment (vdemeester) [inline opts/opts.go]: I'm also wondering if we should modify this in this PR or on another one, just for that (so it's easier to review the change and easier to revert too). @duglin WDYT ?

Comment (duglin) [inline opts/opts.go]: I was the one that added that comment, not because the func shouldn't validate the mode part, but rather to make sure people knew that it didn't.  I actually would prefer if it did. If you want to do that as part of a different PR that would be fine with me, but we probably need to do the other part I mentioned in the previous comment.

Comment (vdemeester) [inline opts/opts.go]: > I was the one that added that comment, not because the func shouldn't validate the mode part, but rather to make sure people knew that it didn't. I actually would prefer if it did.

Go it and I do agree :wink:.

> If you want to do that as part of a different PR that would be fine with me, but we probably need to do the other part I mentioned in the previous comment.

I'm not sure of which comment you're refering too :blush: :sweat_smile:.

Comment (duglin) [inline opts/opts.go]: I just mean I think we should make sure we parse "/rw:/ro" correctly in this PR.

File: opts/opts_test.go (hunk 29 above)
Comment (vdemeester) [inline opts/opts_test.go]: Currently :
```
/path:rw   # does fail, containerPath should be absolute
path:ro   # does fail because path isn't abs
/rw:rw   # does fail, containerPath should be absolute
/rw:/ro   # should work with /rw as the host path and /ro as the container path
```

Volumes are ``hostPath:containerPath:mode`` and it's the containerPath that needs to be absolute, so you probably meant

```
rw:/path   # should work
path:ro   # should fail because path isn't abs
rw:/rw   # should work with /rw as the container path
/rw:/ro   # should work with /rw as the host path and /ro as the container path
```

File: opts/envfile.go (hunk 26 above)
Comment (vdemeester) [inline opts/envfile.go]: This is done.

File: opts/envfile.go [hunk 31]
```diff
@@ -4,9 +4,16 @@ import (
...
 	"fmt"
 	"os"
+	"regexp"
 	"strings"
 )
 
+var (
+	// Variables set by the user must have a name consisting solely of
+	// alphabetics, numerics, and underscores - the first of which must not be numeric.
+	EnvironmentVariableRegexp = regexp.MustCompile("^[[:alpha:]_][[:alpha:][:digit:]_]+$")
```
Comment (duglin) [inline opts/envfile.go]: I think the "+" needs to be "*" otherwise we won't support single-char env names.

Comment (duglin) [inline opts/envfile.go]: Might be good to add a test for that too if we don't already have one.

Comment (vdemeester) [inline opts/envfile.go]: Ohh your right :blush: I missed that.

File: opts/opts_test.go (hunk 29 above)
Comment (duglin) [inline opts/opts_test.go]: Well, this get to the fun part of this.  :-)  Looking at `/path:rw`... is it meant to be hostPath:containerPath or containerPath:mode ?  IMO `/path:rw` should work because its valid per the syntax and it should be containerPath:mode.  I don't think we can assume that if there's just one colon then the user meant hostPath:containerPath - I think we may need to try to be smart about it and look at what's after the last colon and see if its a valid mode, and if so then assume its not a containerPath.  And, likewise, if its not a valid mode then assume its hostPath:containerPath.  Not sure what else we can do - its unfortunate that we use the same separator in both places and we have so many optional bits.

Comment (vdemeester) [inline opts/opts_test.go]: @duglin I agree with you, we should try to be a little smart, so I'm all for check for ro/rw if we just have one ``:``. I'll prepare a separate PR for that.

> Not sure what else we can do - its unfortunate that we use the same separator in both places and we have so many optional bits.

I guess ``:`` is pretty common for this kind of thing. I think having a different char to separate ``hostPath`` and ``containerPath`` and ``mode`` would feel weird in a user point of view.

Comment (vdemeester) [inline opts/opts_test.go]: @duglin This is kinda how it will look : https://github.com/vdemeester/docker/commit/b3c0d6ca62260d2eff74f2fd1623c4f4fd8bb6dc but I still think it's better in another PR (or in a separate commit maybe ?)

**Comment by vdemeester** [comment]
@duglin I've put both commit on this. Tell me if you'ld prefer that I squash it to one commit or not :wink:.

File: opts/opts.go [hunk 32]
```diff
@@ -197,20 +197,36 @@ func ValidateLink(val string) (string, error) {
...
+		return val, fmt.Errorf("bad format for volumes: %s", val)
+	}
+	switch len(splited) {
+	case 1:
 		containerPath = splited[0]
-		val = path.Clean(splited[0])
-	} else {
+		val = path.Clean(containerPath)
+	case 2:
+		if splited[1] == "rw" || splited[1] == "ro" {
```
Comment (duglin) [inline opts/opts.go]: I think we may need to externalize and then use validMountMode() from daemon/volumes.go since the list of modes is more than just rw and ro.  What do you think?

**Comment by duglin** [comment]
Just one question/comment! Looking good!

File: opts/opts.go (hunk 32 above)
Comment (vdemeester) [inline opts/opts.go]: You're right :smile_cat:. I'll look at that this evening.

File: runconfig/parse_test.go [hunk 33]
```diff
@@ -157,9 +161,6 @@ func TestParseRunVolumes(t *testing.T) {
 	if _, _, err := parse(t, "-v /tmp:"); err == nil {
 		t.Fatalf("Error parsing volume flags, `-v /tmp:` should fail but didn't")
 	}
-	if _, _, err := parse(t, "-v /tmp:ro"); err == nil {
```
Comment (vdemeester) [inline runconfig/parse_test.go]: This one will not fail anymore, so removing it :)

**Comment by vdemeester** [comment]
@duglin I think it's done.. I'm not found of the move I made of ``rwMode`` and ``roMode``. WDYT ? :sweat_smile:

**Comment by duglin** [comment]
@vdemeester yea, lets keep that code into volumes.go and just externalize validateMountMode() instead of moving it all into opt.go. To me, volumes (sort of) owns all that logic and the opts stuff is more like a user of it.

**Comment by vdemeester** [comment]
@duglin The problem is there seems to be a *cycle* loop of import (if I import ``daemon`` in the package ``opts``) . I'm gonna move it into ``github.com/docker/docker/volume``.

File: runconfig/parse_test.go (hunk 33 above)
Comment (vdemeester) [inline runconfig/parse_test.go]: Meaning, now it is supposed to work with this syntax and not fail like on this test :wink:.

**Comment by vdemeester** [comment]
@duglin quick update :
1. Moved ``ValidateMountMode`` to ``github.com/docker/docker/volume``.
2. Fixed few tests (``runconfig``, etc..), e.g. that would failed because they were considered ``rwz`` that would be later rejected as not supported.
3. Added a bunch of *godoc* :wink:.

**Comment by vdemeester** [comment]
@duglin Another update :sweat_smile:.
I've added a ``ValidateDevice`` function that will do the same as ``ValidatePath`` but without validating the mount mode (https://github.com/docker/docker/pull/13694/files#diff-def63b7f78492fa1f0dc4e219a650df4R229). It should fix the ``DockerSuite.TestDevicePermissions`` integration test.

**Comment by duglin** [comment]
@vdemeester thanks for the hard work!
LGTM

ping @LK4D4 you still ok with this?

**Comment by calavera** [comment]
LGTM

**Comment by thaJeztah** [comment]
:tada: thanks!!

**Comment by vdemeester** [comment]
:tada: :blush: 
@duglin thanks for all the reviews !

**Comment by thaJeztah** [comment]
Added impact labels, because the validation-rules for environment variables changed in this PR (see https://github.com/docker/docker/pull/14727)
